Source/Programs/wavm.cpp

```diff
@@ -0,0 +1,224 @@
+			sourceFile = *args;
+		}
+		else if(!sourceFile && endsWith(*args, ".wast"))
+		{
+			sourceFile = *args;
+		}
+		else if(!binaryFile && !strcmp(*args, "--binary"))
+		{
+			if (!*++args) { showHelp(); return EXIT_FAILURE; }
+			binaryFile = *args;
+			if (!*++args) { showHelp(); return EXIT_FAILURE; }
+			binaryMemFile = *args;
+		}
+		else if(!binaryFile && endsWith(*args, ".wasm"))
+		{
+			binaryFile = *args;
+			if (!*++args) { showHelp(); return EXIT_FAILURE; }
+			binaryMemFile = *args;
+		}
+		else if(!strcmp(*args, "--function") || !strcmp(*args, "-f"))
+		{
+			if (!*++args) { showHelp(); return EXIT_FAILURE; }
+			functionName = *args;
+		}
+		else if(!strcmp(*args, "--outputAST"))
+		{
+			if (!*++args) { showHelp(); return EXIT_FAILURE; }
+			outputASTFilename = *args;
+		}
+		else if(!strcmp(*args, "--check") || !strcmp(*args, "-c"))
+		{
+			onlyValidate = true;
+		}
+		else if(!strcmp(*args, "--"))
+		{
+			++args;
+			break;
+		}
+		else if(!strcmp(*args, "--help") || !strcmp(*args, "-h"))
+		{
+			showHelp();
+			return EXIT_SUCCESS;
+		}
+		else {
+			break;
+		}
+	}
+
+	AST::Module* module = nullptr;
+	const char* main_arg0 = 0;
+	if (sourceFile)
+	{
+		WebAssemblyText::File wastFile;
+		if(loadTextModule(sourceFile,wastFile)) { module = wastFile.modules[0]; }
+		else { return EXIT_FAILURE; }
+		main_arg0 = sourceFile;
+	}
+	else if (binaryFile)
+	{
+		module = loadBinaryModule(binaryFile,binaryMemFile);
+		main_arg0 = binaryFile;
+	}
+	else
+	{
+		showHelp();
+		return EXIT_FAILURE;
+	}
+
+	if(!module) { return EXIT_FAILURE; }
+	if(!Runtime::init()) { return EXIT_FAILURE; }
+	if(!Runtime::loadModule(module)) { return EXIT_FAILURE; }
+
+	auto functionExport = module->exportNameToFunctionIndexMap.find(functionName);
+	// Support both main and _main by default
+	if(functionExport == module->exportNameToFunctionIndexMap.end() && !strncmp(functionName, "main", 4)) { functionExport = module->exportNameToFunctionIndexMap.find("_main"); }
+
+	if(functionExport == module->exportNameToFunctionIndexMap.end())
+	{
+		std::cerr << "Module does not export '" << functionName << "'" << std::endl;
+		return EXIT_FAILURE;
+	}
+
+	if(onlyValidate) { return EXIT_SUCCESS; }
+
+	if(outputASTFilename)
+	{
+		#if WAVM_TIMER_OUTPUT
+		Core::Timer printTimer;
+		#endif
+		std::ofstream outputStream(outputASTFilename);
+		if(!outputStream.is_open())
+		{
+			std::cerr << "Failed to open " << outputASTFilename << ": " << std::strerror(errno) << std::endl;
+			return EXIT_FAILURE;
+		}
+		outputStream << WebAssemblyText::print(module);
+		outputStream.close();
+		#if WAVM_TIMER_OUTPUT
+		printTimer.stop();
+		std::cout << "Printed WAST code in " << printTimer.getMilliseconds() << "ms" << std::endl;
+		#endif
+		return EXIT_SUCCESS;
+	}
+
+	Runtime::Value *parameters = nullptr;
+	auto function = module->functions[functionExport->second];
+	if(!strcmp(functionName, "main") || !strcmp(functionName, "_main"))
```

Comment:
I think this would be more intuitive for users if it instead took this branch based on whether explicit `--function` argument was present or not.